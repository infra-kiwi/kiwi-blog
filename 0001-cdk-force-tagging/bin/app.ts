#!/usr/bin/env node

/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import 'source-map-support/register';
import * as cdk from 'aws-cdk-lib';
import { KiwiStack } from '../lib/stack';

import { name as projectName } from '../package.json';
import { getSynthesizer } from '../lib/synthesizer';

// https://docs.aws.amazon.com/cdk/v2/guide/environments.html
// Use the current environment variables to figure out which account
// and region we want to deploy to
const env: cdk.Environment = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION
};

const app = new cdk.App();

async function main() {
  new KiwiStack(app, projectName, {
    env,

    // Use our custom tagging synthesizer
    synthesizer: await getSynthesizer()
  });
}

void main();
