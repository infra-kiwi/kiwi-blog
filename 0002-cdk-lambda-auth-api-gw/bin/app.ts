#!/usr/bin/env node
import * as cdk from 'aws-cdk-lib';
import { aws_iam } from 'aws-cdk-lib';

import { name as projectName } from '../package.json';
import { ServerStack } from '../lib/stacks/server-stack';
import { ClientStack } from '../lib/stacks/client-stack';

const app = new cdk.App();

// We expect this variable to be provided at deploy time,
// which will decide if we are deploying the server stack
// or the client stack
const component = app.node.getContext('component');

// Use the current environment variables to figure out which account
// and region we want to deploy to
const env = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION
};

function renderServer() {
  // We can, for example, either authorize an account or even the
  // entire AWS organization to make calls to our route
  const clientAccountId = app.node.tryGetContext('clientAccountId');
  // The unique identifier (ID) of an organization (i. e. o-12345abcde)
  const organizationId = app.node.tryGetContext('organizationId');

  // We build the list of allowed principals, which will
  // be able to make requests to the API Gateway route
  const allowedPrincipals: aws_iam.IPrincipal[] = [];
  if (clientAccountId) {
    allowedPrincipals.push(new aws_iam.AccountPrincipal(clientAccountId));
  } else if (organizationId) {
    allowedPrincipals.push(new aws_iam.OrganizationPrincipal(organizationId));
  } else {
    throw new Error(
      'You need to provide either clientAccountId or organizationId'
    );
  }

  new ServerStack(app, projectName + '-server', {
    env,
    allowedPrincipals
  });
}

function renderClient() {
  // Get the server gateway configuration. These values will be
  // the ones returned in the CloudFormation outputs of the Server stack
  const serverGatewayUrl = app.node.getContext('serverGatewayUrl');
  const serverGatewayRegion = app.node.getContext('serverGatewayRegion');

  new ClientStack(app, projectName + '-client', {
    env,
    serverGatewayUrl,
    serverGatewayRegion
  });
}

if (component === 'server') {
  renderServer();
} else if (component === 'client') {
  renderClient();
} else {
  throw new Error(`Unknown component: ${component}`);
}
