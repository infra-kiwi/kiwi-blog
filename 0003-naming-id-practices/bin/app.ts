#!/usr/bin/env node

/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';

import { name as projectName } from '../package.json';
import { StackGoodName } from '../lib/stack-good-name';
import { StackBadName } from '../lib/stack-bad-name';

// https://docs.aws.amazon.com/cdk/v2/guide/environments.html
// Use the current environment variables to figure out which account
// and region we want to deploy to
const env = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION
};

const app = new cdk.App();

new StackGoodName(app, projectName + '-good-name', { env });

new StackBadName(app, projectName + '-bad-name', { env });
