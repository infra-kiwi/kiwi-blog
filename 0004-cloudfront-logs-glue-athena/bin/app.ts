#!/usr/bin/env node
import * as cdk from 'aws-cdk-lib';
import { CloudFrontStack } from '../lib/stack';

import { name as projectName } from '../package.json';

const app = new cdk.App();
new CloudFrontStack(app, projectName, {
  // Use the current environment variables to figure out which account
  // and region we want to deploy to
  env: {
    account: process.env.CDK_DEFAULT_ACCOUNT,
    region: process.env.CDK_DEFAULT_REGION
  }
});
