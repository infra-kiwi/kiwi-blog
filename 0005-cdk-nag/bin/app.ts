#!/usr/bin/env node
import * as cdk from 'aws-cdk-lib';
import { Aspects } from 'aws-cdk-lib';
import { KiwiStack } from '../lib/stack';

import { name as projectName } from '../package.json';
import { AwsSolutionsChecks } from 'cdk-nag';

// https://docs.aws.amazon.com/cdk/v2/guide/environments.html
// Use the current environment variables to figure out which account
// and region we want to deploy to
const env: cdk.Environment = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION
};

const app = new cdk.App();

// Check for bad CDK practices
Aspects.of(app).add(new AwsSolutionsChecks());

new KiwiStack(app, projectName, { env });
