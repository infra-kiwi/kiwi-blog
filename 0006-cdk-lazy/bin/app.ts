#!/usr/bin/env node

/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { KiwiStack } from '../lib/stack';

import { name as projectName } from '../package.json';

// https://docs.aws.amazon.com/cdk/v2/guide/environments.html
// Use the current environment variables to figure out which account
// and region we want to deploy to
const env: cdk.Environment = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION
};

const app = new cdk.App();
new KiwiStack(app, projectName, { env });
