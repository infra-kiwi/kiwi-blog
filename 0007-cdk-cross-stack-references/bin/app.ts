/*
 * Copyright (c) 2024-2025. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { SourceStackWithOutputs } from '../lib/source-stack-with-outputs';
import { DestinationStack } from '../lib/destination-stack';
import { SourceStackWithManualNames } from '../lib/source-stack-with-manual-names';
import { SourceStackWithParameters } from '../lib/source-stack-with-parameters';

import { name as projectName } from '../package.json';
import { App } from 'aws-cdk-lib';

// https://docs.aws.amazon.com/cdk/v2/guide/environments.html
// Use the current environment variables to figure out which account
// and region we want to deploy to
const env: cdk.Environment = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION
};

export function main(app: App) {
  // Define a set of stacks that use automatically generated outputs
  const sourceStackWithOutputs = new SourceStackWithOutputs(
    app,
    `${projectName}-SourceStackWithOutputs`,
    { env }
  );
  const destinationStackFromOutputs = new DestinationStack(
    app,
    `${projectName}-DestinationStackFromOutputs`,
    {
      env,
      snsTopicForAlarms: sourceStackWithOutputs.topic
    }
  );

  // Define a set of stacks that use manually generated resource names
  const sourceStackWithManualNames = new SourceStackWithManualNames(
    app,
    `${projectName}-SourceStackWithManualNames`,
    { env }
  );
  const destinationStackFromManualNames = new DestinationStack(
    app,
    `${projectName}-DestinationStackFromManualNames`,
    {
      env,
      snsTopicForAlarmsName: sourceStackWithManualNames.topicName
    }
  );
  // Force DestinationStackFromManualNames to be created
  // AFTER SourceStackWithManualNames
  destinationStackFromManualNames.addDependency(sourceStackWithManualNames);

  // Define a set of stacks that use SSM parameters
  const sourceStackWithParameters = new SourceStackWithParameters(
    app,
    `${projectName}-SourceStackWithParameters`,
    { env }
  );
  const destinationStackFromParameters = new DestinationStack(
    app,
    `${projectName}-DestinationStackFromParameters`,
    {
      env,
      snsTopicForAlarmsParameterName: sourceStackWithParameters.parameterName
    }
  );
  // Force DestinationStackFromParameters to be created
  // AFTER SourceStackWithParameters
  destinationStackFromParameters.addDependency(sourceStackWithParameters);

  return {
    sourceStackWithOutputs,
    destinationStackFromOutputs,
    sourceStackWithManualNames,
    destinationStackFromManualNames,
    sourceStackWithParameters,
    destinationStackFromParameters
  };
}

if (require.main == module) {
  void main(new cdk.App());
}
