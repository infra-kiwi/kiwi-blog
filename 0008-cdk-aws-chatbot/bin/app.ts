/*
 * Copyright (c) 2024-2025. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { ChatBotStack } from '../lib/stack';

import { name as projectName } from '../package.json';
import { loadConfig } from '../lib/util/config';

// https://docs.aws.amazon.com/cdk/v2/guide/environments.html
// Use the current environment variables to figure out which account
// and region we want to deploy to
const env: cdk.Environment = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION
};

const app = new cdk.App();

// Accepts the `-c configFile=my-config.yml` parameter, or
// defaults to a common file name
const configFile = app.node.tryGetContext('configFile') ?? 'config.yml';

async function createStack(app: cdk.App) {
  const config = await loadConfig(configFile);

  new ChatBotStack(app, projectName, {
    env,
    config
  });
}

void createStack(app);
