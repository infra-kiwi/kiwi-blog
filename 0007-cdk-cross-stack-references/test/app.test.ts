/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { main } from '../bin/app';

describe('Stacks', () => {
  const app = new App();

  const stacks = main(app);

  test('Outputs', () => {
    const template = Template.fromStack(stacks.sourceStackWithOutputs);
    expect(template).toMatchSnapshot();

    // Verify the stack contains a generated output
    template.hasOutput('*', {
      Value: stacks.sourceStackWithOutputs.resolve(
        stacks.sourceStackWithOutputs.topic.topicArn
      )
    });

    const outputs = template.toJSON()['Outputs'];
    expect(Object.keys(outputs).length).toEqual(1);
  });

  test('Hardcoded names', () => {
    const template = Template.fromStack(stacks.sourceStackWithManualNames);
    expect(template).toMatchSnapshot();

    // Verify there are no outputs
    expect(template.toJSON()['Outputs']).toEqual(undefined);
  });

  test('SSM parameters', () => {
    const template = Template.fromStack(stacks.sourceStackWithParameters);
    expect(template).toMatchSnapshot();

    // Verify the stack contains an SSM parameter that points to the source SNS topic
    template.resourcePropertiesCountIs(
      'AWS::SSM::Parameter',
      {
        Value: stacks.sourceStackWithParameters.resolve(
          stacks.sourceStackWithParameters.topic.topicArn
        )
      },
      1
    );

    // Verify there are no outputs
    expect(template.toJSON()['Outputs']).toEqual(undefined);
  });
});
