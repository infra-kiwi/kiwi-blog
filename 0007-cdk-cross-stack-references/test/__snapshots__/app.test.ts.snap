// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stacks Hardcoded names 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "TopicWithAName61936DD2": {
      "Properties": {
        "TopicName": "kiwi-blog-0007-cdk-cross-stack-references-SourceStackWithManualNames-topic-with-a-name",
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Stacks Outputs 1`] = `
{
  "Outputs": {
    "ExportsOutputRefMyTopic868694349D03D60F": {
      "Export": {
        "Name": "kiwi-blog-0007-cdk-cross-stack-references-SourceStackWithOutputs:ExportsOutputRefMyTopic868694349D03D60F",
      },
      "Value": {
        "Ref": "MyTopic86869434",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "MyTopic86869434": {
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Stacks SSM parameters 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "TopicBFC7AF6E": {
      "Type": "AWS::SNS::Topic",
    },
    "TopicForAlarms6F9A7238": {
      "Properties": {
        "Name": "kiwi-blog-0007-cdk-cross-stack-references-SourceStackWithParameters-topic-arn",
        "Type": "String",
        "Value": {
          "Ref": "TopicBFC7AF6E",
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
